mapper-runtime/src/main/java/com/datastax/oss/driver/internal/mapper/DefaultMapperContext.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.mapper;
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.mapper.MapperContext;
+import com.datastax.oss.driver.api.mapper.entity.naming.NameConverter;
+import com.datastax.oss.protocol.internal.util.collection.NullAllowingImmutableMap;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class DefaultMapperContext implements MapperContext {
+
+  private final CqlSession session;
+  private final CqlIdentifier keyspaceId;
+  private final CqlIdentifier tableId;
+  private final ConcurrentMap<Class<? extends NameConverter>, NameConverter> nameConverterCache;
+  private final Map<Object, Object> customState;
+
+  public DefaultMapperContext(
+      @NonNull CqlSession session, @NonNull Map<Object, Object> customState) {
+    this(
+        session,
+        null,
+        null,
+        new ConcurrentHashMap<>(),
+        NullAllowingImmutableMap.copyOf(customState));
+  }
+
+  private DefaultMapperContext(
+      CqlSession session,
+      CqlIdentifier keyspaceId,
+      CqlIdentifier tableId,
+      ConcurrentMap<Class<? extends NameConverter>, NameConverter> nameConverterCache,
+      Map<Object, Object> customState) {
+    this.session = session;
+    this.keyspaceId = keyspaceId;
+    this.tableId = tableId;
+    this.nameConverterCache = nameConverterCache;
+    this.customState = customState;
+  }
+
+  public DefaultMapperContext withKeyspaceAndTable(
+      @Nullable CqlIdentifier newKeyspaceId, @Nullable CqlIdentifier newTableId) {
+    return (Objects.equals(newKeyspaceId, this.keyspaceId)
```

Comment:
What is the reason for that logic? (returning `this`)
Do we just want to use `withKeyspaceAndTable` to achieve `computeIfAbsent` semantic?

Reply:
`DefaultMapperContext` is immutable, calling `withKeyspaceAndTable` returns a new instance.

It's more convenient that way, because every DAO factory method uses its own keyspace/table combination, and those are called concurrently.

The only part that is not mutable is the name converter cache, but name converters are global to a mapper instance.